### ISSUE TITLE
optimization flags for autotools packages

### ISSUE THREAD

**Issue by mwkrentel** [spack/spack#14430]
Spack and autotools don't play nice when it comes to debug and
optimization flags.

I remember when I first started writing spack packages, I wanted the
recipes to respect cflags.  That is, I should be able to write
`cflags='-O1'` or `-O3` on the spack install line, with or without `-g`
and have that correctly passed through to autotools configure.

This is actually rather awkward in spack, largely due to how autotools
configure treats CFLAGS.  If you omit CFLAGS from the configure line,
autotools chooses a default (normally, `-g -O2`) for you, but if you
specify CFLAGS, then it uses exactly that.  If you want `-O2`, you
have to write it yourself.

Note: autotools is the wild west of build systems.  Any package is
free to make its own rules.  But here, let's mainly consider the
common behavior of passing flags to the configure script via CFLAGS
and selecting `-g -O2` by default.

Let's consider different ways of passing spack's cflags to configure.

1. The default `inject_flags`.  With inject_flags, spack adds my
cflags to the compiler wrapper (along with the microarch flags).  The
problem is that configure sees CFLAGS as unspecified and chooses its
own default.  For example:

```
spack -d install -v libsigsegv cflags='-O1'
...
[cc] /bin/gcc -march=westmere -mtune=westmere -O1 -I. -I.. -I. -I.
   -DHAVE_CONFIG_H -g -O2 -c handler.c -o handler.o
```

So, I can write `cflags='-O0'` or `-O9` and configure resets this to
`-g -O2`.  That's fine for a default, but the goal is to obey what I
tell spack, and this doesn't.

And don't get the idea that, "We'll just move the flags to the end of
the line so that the spack flags win."  I know packages (libmonitor)
that need to see CFLAGS and adjust them to make some code compile
(involving asm and code motion).  This idea would break those
packages.

Btw, long ago, the situation was much worse.  Before `flag_handler`,
spack would pass `CFLAGS=''` to configure and a lot of packages would
get built unoptimized.  In autotools, there is a huge difference
between an unspecified CFLAGS and CFLAGS=''.

2. `build_system_flags`.  This passes CFLAGS to configure.  So, if I
say `cflags='-O1'`, then I get exactly that.

The problem here is that if you specify cflags, any cflags, then you
need to specify everything.  If you write `cflags='-mavx'`, then
that's all I get, no `-O2`.  This is awkward and error prone.

Plus, I've never really liked that spack's cflags are passed down to
dependencies.  I'd prefer a debug/opt option that applies to this
package only.

IMO, I think the takeaway is: autotools expects to see CFLAGS, and so
it's better to take the decision out of configure and push it up to
the spack level.

Note: this already exists for CMakePackage.  Every cmake package has a
`build_type` variant, so this is cleaner for cmake.

----------

I propose that we add variants for `debug_symbols` (boolean) and
`opt_flags` (string) to the AutotoolsPackage class (maybe all
packages), make the decision at the spack level, use this to set
CFLAGS, CXXFLAGS and FFLAGS and pass these to configure.

If the user also specifies `cflags`, then add those to the debug and
opt flags.  I'm suggesting that we factor out the debug and opt flags
because that's what configure sets on its own.

Actually, you can't impose this on all autotools packages, that would
surely break some packages.  Make it an easy, 1-knob option for the
packages that can use it.

Btw, `debug` is a bit ambiguous, does it mean low optimization (-O0)
or debug symbols (-g) or both.  I went with `debug_symbols` to be more
explicit (plus many packages already have a `debug` variant).

Also, I went with `opt_flags` (string) instead of `opt_level`
(integer) to allow for things like `-Os` or `-Ofast`.

Fundamentally, I think it makes sense to decide this at the spack
level and then pass that to configure.  Cmake already does this and I
think the other package classses should make this available.

@tgamblin

**Comment by haampie** [comment]
This issue should really get some attention. There's over 900 autotools packages, and the current situation is just really cumbersome.

One would expect `spack install [pkg] cflags='-O3'` to force the O3 flag, but it almost never happens:

```
 $ spack spec -I binutils cflags='-O3' 
Input spec
--------------------------------
 -   binutils cflags="-O3" 

Concretized
--------------------------------
[+]  binutils@2.34%gcc@9.3.0 cflags="-O3" +gold~headers~interwork~ld~libiberty~lto+nls~plugins arch=linux-ubuntu20.04-zen2
[+]      ^gettext@0.20.2%gcc@9.3.0 cflags="-O3" +bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-zen2
[+]          ^bzip2@1.0.8%gcc@9.3.0 cflags="-O3" +shared arch=linux-ubuntu20.04-zen2
[+]              ^diffutils@3.7%gcc@9.3.0 cflags="-O3"  arch=linux-ubuntu20.04-zen2
[+]                  ^libiconv@1.16%gcc@9.3.0 cflags="-O3"  arch=linux-ubuntu20.04-zen2
[+]          ^libxml2@2.9.10%gcc@9.3.0 cflags="-O3" ~python arch=linux-ubuntu20.04-zen2
[+]              ^pkgconf@1.7.3%gcc@9.3.0 cflags="-O3"  arch=linux-ubuntu20.04-zen2
[+]              ^xz@5.2.5%gcc@9.3.0 cflags="-O3" ~pic arch=linux-ubuntu20.04-zen2
[+]              ^zlib@1.2.11%gcc@9.3.0 cflags="-O3" +optimize+pic+shared arch=linux-ubuntu20.04-zen2
[+]          ^ncurses@6.2%gcc@9.3.0 cflags="-O3" ~symlinks+termlib arch=linux-ubuntu20.04-zen2
[+]          ^tar@1.32%gcc@9.3.0 cflags="-O3"  arch=linux-ubuntu20.04-zen2
[+]      ^texinfo@6.5%gcc@9.3.0 cflags="-O3"  patches=12f6edb0c6b270b8c8dba2ce17998c580db01182d871ee32b7b6e4129bd1d23a,1732115f651cff98989cb0215d8f64da5e0f7911ebf0c13b064920f088f2ffe1 arch=linux-ubuntu20.04-zen2
[+]          ^perl@5.30.3%gcc@9.3.0 cflags="-O3" +cpanm+shared+threads arch=linux-ubuntu20.04-zen2
[+]              ^berkeley-db@18.1.40%gcc@9.3.0 cflags="-O3"  arch=linux-ubuntu20.04-zen2
[+]              ^gdbm@1.18.1%gcc@9.3.0 cflags="-O3"  arch=linux-ubuntu20.04-zen2
[+]                  ^readline@8.0%gcc@9.3.0 cflags="-O3"  arch=linux-ubuntu20.04-zen2
```

Looking through the build logs, I find the following effective compiler flags:

```
berkeley : -O3
binutils : -g -O2
bzip2    : -O2 -g
gettext  : -g -O2
libiconv : -g -O2
ncurses  : -O2
perl     : -O2
readline : -g -O
texinfo  : -g -O2
xz       : -g -O2
zlib     : -O2
```
Only `berkeley` seems to respect it.

---

@mwkrentel: Can't we take approach 2 with `build_system_flags`, where we just go through the a bunch of autotools packages and copy the default flags into `package.py`, and conditionally add optimization flags when the user has not provided them?

Basically:

```
def flag_handler(self, name, flags):
  opt_flag_found = any(f in self.compiler.opt_flags for f in flags)

  if name == 'cflags':
    # maybe add some defaults here...
    if not opt_flag_found:
      flags.append('-O2') # the default optimization flag in the package

  return (None, None, flags)
```

This way the defaults make sense, and they can be overridden with `spack install [spec] cflags=...`

**Comment by haampie** [comment]
Ok, after a chat with @alalazo: the list of optimization flags is copied from the build logs, but does not take into account what the spack compiler wrapper actually does with it. So, it could be that the compiler wrapper swallows the `-O2` flag and actually turns it into `-O3`. Should be verified

**Comment by mwkrentel** [comment]
@scheibelp I see where you self-assigned this recently.

Are you thinking of writing an implementation for this, possibly
building on #16710 ?  If yes, let me know.  I'd like to be involved in
the design.

My basic point is that the options for debug and optimization flags
should be pushed out to the spack level, similar to how `build_type`
is used for cmake.

**Comment by RemiLacroix-IDRIS** [comment]
> Ok, after a chat with @alalazo: the list of optimization flags is copied from the build logs, but does not take into account what the spack compiler wrapper actually does with it. So, it could be that the compiler wrapper swallows the `-O2` flag and actually turns it into `-O3`. Should be verified

The compiler flags passed via `spack install` are indeed injected through the compiler wrapper but I think that anything on the command-line will actually overwrite those, at least with most compilers.

You will end up with something like `<compiler> <injected flags> <flags from the actual command-line>` and all the compilers I know will consider only the last occurrence when an option is given multiple times (i.e. `gcc -O3 -O2 ...` is the same as `gcc -O2 ...`.

**Comment by haampie** [comment]
Yeah, I came to that conclusion too. You can potentially patch the compiler wrapper

```diff
--- a/lib/spack/env/cc
+++ b/lib/spack/env/cc
@@ -368,6 +368,12 @@ while [ -n "$1" ]; do
                 other_args+=("$1")
             fi
             ;;
+        -O|-O0|-O1|-O2|-O3|-O4|-Ofast|-Os|-Oz|-Og)
+            if [[ -z "$SPACK_FORCED_OPTIMIZATION_FLAGS"  ]]; then
+                other_args+=("$1")
+            else
+                other_args+=("$SPACK_FORCED_OPTIMIZATION_FLAGS")
+            ;;
         *)
             if [[ "$1" = "$dtags_to_strip" ]] ; then
                 :  # We want to remove explicitly this flag
```

and then run `SPACK_FORCED_OPTIMIZATION_FLAGS=-O3 spack install ...`

**Comment by mwkrentel** [comment]
The value of `cflags` is normally passed to the compiler wrapper,
either at the front or end of the compile line depending on the value
of `flag_handler`.  (Theoretically, `flag_handler` could do anything,
erase the flags, completely reset them, but normally you pass then
through.)

If you put them at the front, then autotools won't see them and choose
its own default which will override cflags.

If you put them at the end, then you're lying to autotools about the
real flags and I know this will break at least one package that I
maintain (libmonitor).  (In libmonitor, we need to adjust the flags
for some files that use asm lines.)

Plus, `cflags` is passed down to dependents, which is not always what
you want.

----------

IMO, what you really need to do is decide the value at the spack level
with a variant, and then pass the value to CFLAGS, CXXFLAGS so that
you work in cooperation with autotools and not try to overrule it.

The variant could either be one variable (like `build_flags` for cmake),
or separate values for optimization level and debug (-g) on/off.
Either way, I don't care.

Cmake gets it right, follow what cmake does for autotools.
But take into account that not all autotools are well behaved.
